### PR TITLE
Feature 207 refine user management

### LINKED ISSUE #207: Refine user management UI
As a user I'd like to have the user management/user handling workflow refined.

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #207

### CODE REVIEW

File: src/components/editor/editor-screen-table.tsx [hunk 1]
```diff
@@ -120,13 +120,13 @@ const EditorScreenTable: FC<Props> = ({
...
-      const typeString = version.type === MetaformVersionType.Archived ? formVersionArchived : formVersionDraft;
-      const lastModifier = lastModifiers.find(user => user.id === lastModifierId);
+      const typeString = type === MetaformVersionType.Archived ? formVersionArchived : formVersionDraft;
+      const lastModifier = lastModifiers.find(modifier => modifier.id === lastModifierId);
 
       return {
         id: id!,
         typeString: typeString,
-        type: type,
+        type: type!,
```
Comment (Chi-Chu319) [inline src/components/editor/editor-screen-table.tsx:129]: Using null assertion is non safe

File: src/components/screens/users/users-screen.tsx [hunk 2]
```diff
@@ -31,6 +32,65 @@ const UsersScreen: React.FC = () => {
   const [ selectedMemberGroupId, setSelectedMemberGroupId ] = React.useState<string>();
   const [ addMemberGroupOpen, setAddMemberGroupOpen ] = React.useState<boolean>(false);
   const [ addMemberOpen, setAddMemberOpen ] = React.useState<boolean>(false);
+  const [ editMemberOpen, setEditMemberOpen ] = React.useState<boolean>(false);
+
+  /**
+   * Search users from the API
```
Comment (Chi-Chu319) [inline src/components/screens/users/users-screen.tsx:38]:
```suggestion
   * Searches users from the API
```

File: src/components/users/edit-member-dialog.tsx [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+const EditMemberDialog: React.FC<Props> = ({
+  loading,
+  open,
+  onCancel,
+  setLoading,
+  searchUsers,
+  editUser
+}) => {
+  const [ selectedMetaformUser, setSelectedMetaformUser ] = useState<User | undefined>();
+  const [ selectedCardUser, setSelectedCardUser ] = useState<User | undefined>();
```
Comment (Chi-Chu319) [inline src/components/users/edit-member-dialog.tsx:36]:
```suggestion
  const [ selectedMetaformUser, setSelectedMetaformUser ] = useState<User>();
  const [ selectedCardUser, setSelectedCardUser ] = useState<User>();
```

File: src/components/users/edit-member-dialog.tsx [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+  /**
+   * Gets UPN number from federated users displayname
+   * if user doesn't contain federated user, returns null
+   */
+  const getUpnNumber = () => {
+    if (selectedCardUser) {
+      return selectedCardUser.displayName?.match(/\d/g)!.join("");
+    }
+
+    const federatedUser = selectedMetaformUser!.federatedIdentities?.find(user => user.source === UserFederationSource.Card);
```
Comment (Chi-Chu319) [inline src/components/users/edit-member-dialog.tsx:167]:
```suggestion
    const federatedUser = selectedMetaformUser?.federatedIdentities?.find(user => user.source === UserFederationSource.Card);
```

File: src/components/users/edit-member-dialog.tsx [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  };
+
+  const valid = selectedMetaformUser?.firstName && selectedMetaformUser?.lastName && EmailValidator.validate(selectedMetaformUser.email);
+
+  /**
+   * Renders dialog content
+   */
+  const renderDialogContent = () => (
+    <Stack spacing={ 1 }>
+      <Stack spacing={ 1 } direction="row" alignItems="center">
```
Comment (Chi-Chu319) [inline src/components/users/edit-member-dialog.tsx:183]:
```suggestion
      <Stack
        spacing={ 1 }
        direction="row"
        alignItems="center"
      >
```

File: src/components/users/edit-member-dialog.tsx [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+    <GenericLoaderWrapper loading={ loading }>
+      <Button onClick={ handleEditClick } color="primary" disabled={ !valid }>
+        { strings.userManagementScreen.editMemberDialog.editButton }
+      </Button>
+    </GenericLoaderWrapper>
+  ];
+
+  useEffect(() => {
+    setLinkSwitchChecked(!!selectedMetaformUser?.federatedIdentities?.length ?? false);
+  }, [ selectedMetaformUser ]);
```
Comment (Chi-Chu319) [inline src/components/users/edit-member-dialog.tsx:305]: Always place the useEffect hook call to the top of a component after the other hook calls and the method it calls.

File: src/components/users/edit-member-dialog.tsx [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+        label={ strings.userManagementScreen.editMemberDialog.cardAuthUsersSelectLabel }
+      >
+        { foundCardUsers.map(user => {
+          const federatedUser = user.federatedIdentities?.find(federatedIdentity => federatedIdentity.source === UserFederationSource.Card);
+          return (
+            <MenuItem key={ user.displayName} value={ federatedUser?.userId }>
+              { `${user.firstName} ${user.lastName}` }
+            </MenuItem>);
+        })
+        }
```
Comment (Chi-Chu319) [inline src/components/users/edit-member-dialog.tsx:237]: Move this to a separate method

File: src/components/users/add-member-dialog.tsx [hunk 8]
```diff
@@ -15,120 +19,277 @@ interface Props {
...
 
-  const valid = member.firstName && member.lastName && EmailValidator.validate(member.email);
+  /**
+   * Event handler for search button click
+   */
+  const handleSearchButtonClick = async () => {
+    const users = (await searchUsers(userSearch))
+      .filter(user => user.displayName !== API_ADMIN_USER)
+      .sort((a: User, b: User) => (a.displayName! < b.displayName! ? -1 : 1));
+    // users.sort((a: User, b: User) => (a.displayName! < b.displayName! ? -1 : 1));
```
Comment (Chi-Chu319) [inline src/components/users/add-member-dialog.tsx:123]:
```suggestion
  const handleSearchButtonClick = async () => {
    const users = (await searchUsers(userSearch))
      .filter(user => user.displayName !== API_ADMIN_USER)
      .sort((a: User, b: User) => (a.displayName! < b.displayName! ? -1 : 1));

```

**Review by Chi-Chu319 — CHANGES_REQUESTED** [review]

File: src/components/users/add-member-dialog.tsx (hunk 8 above)
Comment (villejuutila) [reply src/components/users/add-member-dialog.tsx:123]: Ah, legendary... :|

File: src/components/users/add-member-dialog.tsx [hunk 9]
```diff
@@ -15,120 +19,277 @@ interface Props {
...
+        onChange={ handleTextFieldChange }
+      />
+    </Stack>
+  );
+
+  /**
+   * Renders dialog actions
+   */
+  const renderDialogActions = () => [
+    <Button disableElevation variant="contained" onClick={ handleCancelClick } color="secondary" autoFocus>
```
Comment (EVP1989) [inline src/components/users/add-member-dialog.tsx:271]: Should cancel button also be with loader, does it actually "cancel" anything if api call is made?

File: src/components/users/edit-member-dialog.tsx [hunk 10]
```diff
@@ -0,0 +1,321 @@
+import { User, UserFederationSource } from "generated/client";
+import React, { useEffect, useState } from "react";
+import strings from "../../localization/strings";
+import * as EmailValidator from "email-validator";
+import { Button, FormControlLabel, IconButton, InputAdornment, MenuItem, Stack, Switch, TextField } from "@mui/material";
+import ClearIcon from "@mui/icons-material/Clear";
+import UsersScreenDialog from "./users-screen-dialog";
+import GenericLoaderWrapper from "components/generic/generic-loader";
+
+const API_ADMIN_USER = "api-admin";
```
Comment (EVP1989) [inline src/components/users/edit-member-dialog.tsx:10]: Maybe use enum for this?

File: src/components/users/edit-member-dialog.tsx [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+  onCancel: () => void;
+  setLoading: (value: boolean) => void;
+  searchUsers: (search: string) => Promise<User[]>;
+  editUser: (user: User) => Promise<void>;
+}
+
+/**
+ * React component for edit member dialog
+ */
+const EditMemberDialog: React.FC<Props> = ({
```
Comment (EVP1989) [inline src/components/users/edit-member-dialog.tsx:27]: Move FC to react import

File: src/components/users/edit-member-dialog.tsx [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+    setSelectedCardUser(federatedUser);
+  };
+
+  /**
+   * Searches users by text search
+   */
+  const handleSearchButtonClick = async () => {
+    const users = (await searchUsers(userSearch))
+      .filter(user => user.displayName !== API_ADMIN_USER)
+      .sort((a: User, b: User) => (a.displayName! < b.displayName! ? -1 : 1));
```
Comment (EVP1989) [inline src/components/users/edit-member-dialog.tsx:138]:
```suggestion
      .sort((userA: User, userB: User) => userA.displayName! < userB.displayName! ? -1 : 1);
```

File: src/components/screens/users/users-screen.tsx (hunk 2 above)
Comment (villejuutila) [reply src/components/screens/users/users-screen.tsx:38]: Resolved

File: src/components/users/edit-member-dialog.tsx (hunk 3 above)
Comment (villejuutila) [reply src/components/users/edit-member-dialog.tsx:36]: Resolved

File: src/components/users/edit-member-dialog.tsx (hunk 4 above)
Comment (villejuutila) [reply src/components/users/edit-member-dialog.tsx:167]: Resolved

File: src/components/users/edit-member-dialog.tsx (hunk 6 above)
Comment (villejuutila) [reply src/components/users/edit-member-dialog.tsx:305]: Resolved

File: src/components/users/edit-member-dialog.tsx (hunk 7 above)
Comment (EVP1989) [inline src/components/users/edit-member-dialog.tsx:237]: Put this on separate method.

File: src/components/users/users-screen-dialog.tsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+      open={ open }
+      onClose={ onCancel }
+    >
+      <DialogTitle variant="h2">
+        { dialogTitle }
+      </DialogTitle>
+      <DialogContent sx={{ width: 500 }}>
+        <DialogContentText
+          variant="body1"
+          color="#ccc"
```
Comment (EVP1989) [inline src/components/users/users-screen-dialog.tsx:73]: use theme color

File: src/components/users/edit-member-dialog.tsx (hunk 7 above)
Comment (villejuutila) [reply src/components/users/edit-member-dialog.tsx:237]: Resolved

File: src/localization/en.json [hunk 14]
```diff
@@ -89,7 +89,10 @@
       "createMemberGroup": "Error occurred while creating a user group",
       "loadMembers": "Error occurred while loading users",
       "removeMemberNotFound": "Could not remove user from the group. User was not found",
-      "addMemberNotFound": "Could not add user to the group. User was not found"
+      "addMemberNotFound": "Could not add user to the group. User was not found",
+      "loadUsers": "Error occurred while loading users",
+      "createUser": "Error eccorred while creating user",
```
Comment (EVP1989) [inline src/localization/en.json:94]:
```suggestion
      "createUser": "Error occurred while creating user",
```

File: src/localization/en.json [hunk 15]
```diff
@@ -345,18 +348,21 @@
       "group": "Select group",
       "user": "Select user"
     },
+    "editMemberButton": "Muokkaa käyttäjää",
```
Comment (EVP1989) [inline src/localization/en.json:351]:
```suggestion
    "editMemberButton": "Edit user",
```

**Review by EVP1989 — COMMENTED** [review]
Added comments

File: src/components/users/add-member-dialog.tsx (hunk 8 above)
Comment (villejuutila) [reply src/components/users/add-member-dialog.tsx:123]: Resolved

File: src/components/users/add-member-dialog.tsx [hunk 16]
```diff
@@ -1,11 +1,15 @@
...
+import { Button, IconButton, InputAdornment, ListItem, ListItemIcon, MenuItem, Stack, TextField, Tooltip } from "@mui/material";
+import { MetaformMember, MetaformMemberRole, User, UserFederationSource } from "generated/client";
 import * as EmailValidator from "email-validator";
 import GenericLoaderWrapper from "components/generic/generic-loader";
+import LinkIcon from "@mui/icons-material/Link";
+import CreditCardIcon from "@mui/icons-material/CreditCard";
+import ClearIcon from "@mui/icons-material/Clear";
+import UsersScreenDialog from "./users-screen-dialog";
+
+const API_ADMIN_USER = "api-admin";
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/users/add-member-dialog.tsx:12]: No, but I added it to types and resolved.

File: src/components/users/add-member-dialog.tsx [hunk 17]
```diff
@@ -15,120 +19,277 @@ interface Props {
...
+  setLoading: (value: boolean) => void;
+  searchUsers: (search: string) => Promise<User[]>;
+  createUser: (user: User) => Promise<User | undefined>;
 }
 
 /**
  * React component for add member dialog
  */
-const AddMemberGroupDialog: React.FC<Props> = ({
+const AddMemberDialog: React.FC<Props> = ({
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/users/add-member-dialog.tsx:30]: Resolved

File: src/components/users/add-member-dialog.tsx [hunk 18]
```diff
@@ -15,120 +19,277 @@ interface Props {
...
+  const [ selectedUser, setSelectedUser ] = useState<User | undefined>();
+  const [ userSearch, setUserSearch ] = useState<string>("");
+  const [ foundUsers, setFoundUsers ] = useState<User[]>([]);
+
+  /**
+   * Renders correct icon for User selection dialog
+   * 
+   * @param user User
+   */
+  const renderSelectOptionIcon = (user: User) => {
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/users/add-member-dialog.tsx:48]: Resolved

File: src/components/users/edit-member-dialog.tsx (hunk 10 above)
Comment (villejuutila) [reply src/components/users/edit-member-dialog.tsx:10]: Resolved

File: src/components/users/add-member-dialog.tsx (hunk 9 above)
Comment (villejuutila) [reply src/components/users/add-member-dialog.tsx:271]: Resolved

File: src/components/users/edit-member-dialog.tsx (hunk 12 above)
Comment (villejuutila) [reply src/components/users/edit-member-dialog.tsx:138]: Resolved

File: src/components/users/edit-member-dialog.tsx (hunk 7 above)
Comment (villejuutila) [reply src/components/users/edit-member-dialog.tsx:237]: Resolved

File: src/components/editor/editor-screen-table.tsx (hunk 1 above)
Comment (villejuutila) [reply src/components/editor/editor-screen-table.tsx:129]: Resolved as discussed.

File: src/components/users/edit-member-dialog.tsx (hunk 11 above)
Comment (villejuutila) [reply src/components/users/edit-member-dialog.tsx:27]: Resolved

File: src/components/users/users-screen-dialog.tsx (hunk 13 above)
Comment (villejuutila) [reply src/components/users/users-screen-dialog.tsx:73]: Resolved

File: src/localization/en.json (hunk 14 above)
Comment (villejuutila) [reply src/localization/en.json:94]: Resolved

File: src/localization/en.json (hunk 15 above)
Comment (villejuutila) [reply src/localization/en.json:351]: Resolved

File: src/components/users/add-member-dialog.tsx [hunk 19]
```diff
@@ -15,120 +18,279 @@ interface Props {
...
-  const onTextFieldChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
+  const handleTextFieldChange = (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
     const { target: { name, value } } = event;
     const fieldName = name;
-    const updatedMember: any = { ...member };
-    updatedMember[fieldName] = value;
-    setMember(updatedMember);
+    const updatedUser: any = { ...selectedUser };
+    updatedUser[fieldName] = value;
+    setSelectedUser(updatedUser);
```
Comment (Chi-Chu319) [inline src/components/users/add-member-dialog.tsx:62]: Use produce

File: src/components/users/add-member-dialog.tsx [hunk 20]
```diff
@@ -15,120 +18,279 @@ interface Props {
...
+
+  const valid = selectedUser?.firstName && selectedUser?.lastName && EmailValidator.validate(selectedUser?.email);
+
+  /**
+   * Renders dialog content
+   */
+  const renderDialogContent = () => (
+    <Stack spacing={ 1 }>
+      <Stack spacing={ 1 } direction="row" alignItems="center">
+        <Tooltip title="TOOLTIPPI">
```
Comment (EVP1989) [inline src/components/users/add-member-dialog.tsx:189]: Maybe change this to "notYetImplemented"

**Review by EVP1989 — APPROVED** [review]
Approved

**Review by Chi-Chu319 — APPROVED** [review]

File: src/components/users/add-member-dialog.tsx (hunk 20 above)
Comment (villejuutila) [reply src/components/users/add-member-dialog.tsx:189]: Actually that should've been removed all along. That was the tooltip we were testing yesterday.

File: src/components/users/add-member-dialog.tsx (hunk 19 above)
Comment (villejuutila) [reply src/components/users/add-member-dialog.tsx:62]: Resolved.

**Pull request merged**